# **Data Collection**

## Objectives

* Fetch data from Kaggle and save it as raw data.
* Inspect the data and save it under `outputs/datasets/collection`
  
## Inputs

* Kaggle JSON file (`kaggle.json`) as the authentication token to access the kaggle dataset. 
* Raw dataset from kaggle.com for house prices in Ames, Iowa.

## Outputs

* Generate Dataset: `outputs/datasets/collection/house_prices_records.csv`
* Generate Inherited House Dataset: `outputs/datasets/collection/inherited_houses.csv`

---

# Import all modules required in the workbook

In [1]:
import os
import pandas as pd

# Change working directory

The notebooks for this project are stored in a subfolder called `jupyter_notebooks`, therefore when running the notebook, the working directory needs to be changed to the parent folder. 
* We access the current directory with `os.getcwd()`

In [2]:
current_dir = os.getcwd()
current_dir

'/workspace/pp5-house-price-predictor/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [3]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [4]:
current_dir = os.getcwd()
current_dir

'/workspace/pp5-house-price-predictor'

---

# Fetch Data from Kaggle

Use the `%pip` magic command to install the kaggle package

In [5]:
%pip install kaggle==1.5.12

Note: you may need to restart the kernel to use updated packages.


<div style="border-left: 4px solid #ffa500; padding: 10px; background-color: #f9f9f9; color: #000">
<strong>IMPORTANT:</strong> Remember to import your <code style="color: red">kaggle.json</code> file into the root directory, as this contains the API key.
</div>

Once the `kaggle.json` file is in your root directory, run the cell below, so the token is recognized in the session

In [6]:
os.environ['KAGGLE_CONFIG_DIR'] = os.getcwd()
! chmod 600 kaggle.json

The dataset for this project can be found at [https://www.kaggle.com/datasets/codeinstitute/housing-prices-data](https://www.kaggle.com/datasets/codeinstitute/housing-prices-data)

1. Extract the dataset path from the url. Everything after `www.kaggle.com/datasets/` and assign to a new variable
    - That being: `codeinstitute/housing-prices-data`

2. Define a new variable for the destination folder
3. Download the data from kaggle.

In [7]:
KaggleDatasetPath = "codeinstitute/housing-prices-data"
DestinationFolder = "inputs/datasets/raw"
! kaggle datasets download -d {KaggleDatasetPath} -p {DestinationFolder}

  0%|                                               | 0.00/49.6k [00:00<?, ?B/s]
100%|██████████████████████████████████████| 49.6k/49.6k [00:00<00:00, 1.94MB/s]


Unzip the file, remove the zip file and delete the `kaggle.json` authentication token

In [8]:
! unzip {DestinationFolder}/*.zip -d {DestinationFolder} \
    && rm {DestinationFolder}/*.zip \
    && rm kaggle.json

Archive:  inputs/datasets/raw/housing-prices-data.zip
  inflating: inputs/datasets/raw/house-metadata.txt  
  inflating: inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv  
  inflating: inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv  


---

# Load and Inspect Kaggle data

### Load House Prices Records Data

In [9]:
df = pd.read_csv("inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv")
df.head(10)

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       856     854.0           3.0           No         706          GLQ   
1      1262       0.0           3.0           Gd         978          ALQ   
2       920     866.0           3.0           Mn         486          GLQ   
3       961       NaN           NaN           No         216          ALQ   
4      1145       NaN           4.0           Av         655          GLQ   
5       796     566.0           1.0           No         732          GLQ   
6      1694       0.0           3.0           Av        1369          GLQ   
7      1107     983.0           3.0           Mn         859          ALQ   
8      1022     752.0           2.0           No           0          Unf   
9      1077       0.0           2.0           No         851          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotFrontage  \
0        150            0.0         548          RFn  ...         65.0   
1        284            NaN         460          RFn  ...         80.0   
2        434            0.0         608          RFn  ...         68.0   
3        540            NaN         642          Unf  ...         60.0   
4        490            0.0         836          RFn  ...         84.0   
5         64            NaN         480          Unf  ...         85.0   
6        317            NaN         636          RFn  ...         75.0   
7        216            NaN         484          NaN  ...          NaN   
8        952            NaN         468          Unf  ...         51.0   
9        140            NaN         205          RFn  ...         50.0   

   MasVnrArea OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  WoodDeckSF  \
0       196.0          61            5            7          856         0.0   
1         0.0           0            8            6         1262         NaN   
2       162.0          42            5            7          920         NaN   
3         0.0          35            5            7          756         NaN   
4       350.0          84            5            8         1145         NaN   
5         0.0          30            5            5          796         NaN   
6       186.0          57            5            8         1686         NaN   
7       240.0         204            6            7         1107         NaN   
8         0.0           0            5            7          952         NaN   
9         0.0           4            6            5          991         NaN   

   YearBuilt  YearRemodAdd  SalePrice  
0       2003          2003     208500  
1       1976          1976     181500  
2       2001          2002     223500  
3       1915          1970     140000  
4       2000          2000     250000  
5       1993          1995     143000  
6       2004          2005     307000  
7       1973          1973     200000  
8       1931          1950     129900  
9       1939          1950     118000  

[10 rows x 24 columns]

#### Summary of Data

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       1460 non-null   int64  
 1   2ndFlrSF       1374 non-null   float64
 2   BedroomAbvGr   1361 non-null   float64
 3   BsmtExposure   1460 non-null   object 
 4   BsmtFinSF1     1460 non-null   int64  
 5   BsmtFinType1   1346 non-null   object 
 6   BsmtUnfSF      1460 non-null   int64  
 7   EnclosedPorch  136 non-null    float64
 8   GarageArea     1460 non-null   int64  
 9   GarageFinish   1298 non-null   object 
 10  GarageYrBlt    1379 non-null   float64
 11  GrLivArea      1460 non-null   int64  
 12  KitchenQual    1460 non-null   object 
 13  LotArea        1460 non-null   int64  
 14  LotFrontage    1201 non-null   float64
 15  MasVnrArea     1452 non-null   float64
 16  OpenPorchSF    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  OverallQ

#### Check for Duplicate Data

In [11]:
df[df.duplicated()]

Empty DataFrame
Columns: [1stFlrSF, 2ndFlrSF, BedroomAbvGr, BsmtExposure, BsmtFinSF1, BsmtFinType1, BsmtUnfSF, EnclosedPorch, GarageArea, GarageFinish, GarageYrBlt, GrLivArea, KitchenQual, LotArea, LotFrontage, MasVnrArea, OpenPorchSF, OverallCond, OverallQual, TotalBsmtSF, WoodDeckSF, YearBuilt, YearRemodAdd, SalePrice]
Index: []

[0 rows x 24 columns]

### Load Inherited House Price Data

In [12]:
df_inherited = pd.read_csv("inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv")
df_inherited.head()

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       896         0             2           No       468.0          Rec   
1      1329         0             3           No       923.0          ALQ   
2       928       701             3           No       791.0          GLQ   
3       926       678             3           No       602.0          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotArea  \
0      270.0              0       730.0          Unf  ...    11622   
1      406.0              0       312.0          Unf  ...    14267   
2      137.0              0       482.0          Fin  ...    13830   
3      324.0              0       470.0          Fin  ...     9978   

   LotFrontage MasVnrArea  OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  \
0         80.0        0.0            0            6            5        882.0   
1         81.0      108.0           36            6            6       1329.0   
2         74.0        0.0           34            5            5        928.0   
3         78.0       20.0           36            6            6        926.0   

   WoodDeckSF  YearBuilt  YearRemodAdd  
0         140       1961          1961  
1         393       1958          1958  
2         212       1997          1998  
3         360       1998          1998  

[4 rows x 23 columns]

#### Summary of Data

In [13]:
df_inherited.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       4 non-null      int64  
 1   2ndFlrSF       4 non-null      int64  
 2   BedroomAbvGr   4 non-null      int64  
 3   BsmtExposure   4 non-null      object 
 4   BsmtFinSF1     4 non-null      float64
 5   BsmtFinType1   4 non-null      object 
 6   BsmtUnfSF      4 non-null      float64
 7   EnclosedPorch  4 non-null      int64  
 8   GarageArea     4 non-null      float64
 9   GarageFinish   4 non-null      object 
 10  GarageYrBlt    4 non-null      float64
 11  GrLivArea      4 non-null      int64  
 12  KitchenQual    4 non-null      object 
 13  LotArea        4 non-null      int64  
 14  LotFrontage    4 non-null      float64
 15  MasVnrArea     4 non-null      float64
 16  OpenPorchSF    4 non-null      int64  
 17  OverallCond    4 non-null      int64  
 18  OverallQual   

#### Check for Duplicate Data

In [14]:
df_inherited[df_inherited.duplicated()]

Empty DataFrame
Columns: [1stFlrSF, 2ndFlrSF, BedroomAbvGr, BsmtExposure, BsmtFinSF1, BsmtFinType1, BsmtUnfSF, EnclosedPorch, GarageArea, GarageFinish, GarageYrBlt, GrLivArea, KitchenQual, LotArea, LotFrontage, MasVnrArea, OpenPorchSF, OverallCond, OverallQual, TotalBsmtSF, WoodDeckSF, YearBuilt, YearRemodAdd]
Index: []

[0 rows x 23 columns]

---

# Overall Summary and Observations

**House Price Records Data**
* 1460 records of data
* No duplicated data
* 23 Features and 1 Target variable
* 9 Columns with missing values
* Mixture of Data Types (`int64`, `float64`, `object`)
* 2 Columns with high degree of missing values

**Inherited House data**
* 4 records on the inherited houses data
* No duplicated data
* No missing data
* Mixture of Data Types (`int64`, `float64`, `object`)

**General**
* Manual analysis of the associated text file indicates that some if not all of the object features are ordinal in nature, presenting as textual degrees of quality of a particular feature on the house.
* A consistent approach regarding the data types to be considered if the model requires any performance adjustments. For example, changing int64 to int8/int16 and changing the float64 to int8/int16/int32/64 etc. As some of these data types seem too big to be housing data relating to square feet and/or ordinal numbering. However due to the data size presented, it is unlikely that performance is going to be an issue, but worth noting. 

# Push files to Repo

Create a new folder for the output data and save a copy of the data for further processing. 

In [15]:
try:
    os.makedirs(name='outputs/datasets/collection')
except Exception as e:
    print(e)

df.to_csv('outputs/datasets/collection/house_prices_records.csv', index=False)
df_inherited.to_csv('outputs/datasets/collection/inherited_houses.csv', index=False)

[Errno 17] File exists: 'outputs/datasets/collection'
